src/ray/gcs/gcs_server/gcs_health_check_manager.h

```diff
@@ -0,0 +1,164 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <grpcpp/grpcpp.h>
+
+#include "absl/container/flat_hash_map.h"
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/id.h"
+#include "ray/common/ray_config.h"
+#include "src/proto/grpc/health/v1/health.grpc.pb.h"
+
+class GcsHealthCheckManagerTest;
+
+namespace ray {
+namespace gcs {
+
+/// GcsHealthCheckManager is used to track the healthiness of the nodes in the ray
+/// cluster. The health check is done in pull based way, which means this module will send
+/// health check to the raylets to see whether the raylet is healthy or not. If the raylet
+/// is not healthy for certain times, the module will think the raylet dead.
+/// When the node is dead a callback passed in the constructor will be called and this
+/// node will be removed from GcsHealthCheckManager. The node can be added into this class
+/// later. Although the same node id is not supposed to be reused in ray cluster, this is
+/// not enforced in this class. The implementation of this class try its bes not to couple
+/// itself with GCS so we can move it to ray/common/health_check_manager.h in the future.
+/// TODO (iycheng): Move the GcsHealthCheckManager to common and decouple it from GCS
+/// completely.
+class GcsHealthCheckManager {
+ public:
+  /// Constructor of GcsHealthCheckManager.
+  ///
+  /// \param io_service The thread where all operations in this class should run.
+  /// \param on_node_death_callback The callback function when some node is marked as
+  /// failure.
+  /// \param initial_delay_ms The delay for the first health check.
+  /// \param period_ms The interval between two health checks for the same node.
+  /// \param failure_threshold The threshold before a node will be marked as dead due to
+  /// health check failure.
+  GcsHealthCheckManager(
+      instrumented_io_context &io_service,
+      std::function<void(const NodeID &)> on_node_death_callback,
+      int64_t initial_delay_ms = RayConfig::instance().health_check_initial_delay_ms(),
+      int64_t timeout_ms = RayConfig::instance().health_check_timeout_ms(),
+      int64_t period_ms = RayConfig::instance().health_check_period_ms(),
+      int64_t failure_threshold = RayConfig::instance().health_check_failure_threshold());
+
+  ~GcsHealthCheckManager();
+
+  /// Start to track the healthiness of a node.
+  ///
+  /// \param node_id The id of the node.
+  /// \param channel The gRPC channel to the node.
+  void AddNode(const NodeID &node_id, std::shared_ptr<grpc::Channel> channel);
+
+  /// Stop tracking the healthiness of a node.
+  ///
+  /// \param node_id The id of the node to stop tracking.
+  void RemoveNode(const NodeID &node_id);
+
+  /// Return all the nodes monitored.
```

Comment:
// test-only?

Reply:
I think right now it's only used in testing, but seems like a general API right?